### PR TITLE
Chore/check your answers uses safe page keys

### PR DESCRIPTION
**Pull request by tahb**
This is a proposal to make removing a task from the application more robust. We've run out of time to release this and so need to hand it over.

# Context

[We experienced a production error after removing removing a section from the question form](https://mojdt.slack.com/archives/C05FN6Y2CSV/p1716972264605899). They already had a referral in progress when we released a change to remove a task from the journey, they were then unable to render their check your answers page. We released a fix the user was able to get their existing in progress referral through to submission. 

This would have affected every in progress referral so the potential impact of this bug was much more than 1 user.

This case highlights a fragility in the question framework. When we want to remove a task we have to remember [to hard code that page key into a list of questions to avoid](https://github.com/ministryofjustice/hmpps-community-accommodation-tier-2-ui/blob/60554b22eec1f11d2735d551d95356156bf11415/server/utils/checkYourAnswersUtils.ts#L44). This is easily forgotten as this case demonstrates.

```typescript
pagesKeys.forEach(pageKey => {
    if (!['behaviour-notes', 'behaviour-notes-data', 'reducing-risk'].includes(pageKey)) {
      addPageAnswersToItemsArray({
        items,
        application,
        task,
        pageKey,
        questions,
        outputFormat,
      })
    }
  })
```

N.B when a _question_ is removed, as opposed to a _task_ [this logic is currently protecting us dynamically](https://github.com/ministryofjustice/hmpps-community-accommodation-tier-2-ui/blob/60554b22eec1f11d2735d551d95356156bf11415/server/utils/checkYourAnswersUtils.ts#L93).

As a way to improve this we had the idea to dynamically check the latest tasks against the content of the application. When we find a tasks that were present on creation but are no longer supporter we automatically ignore them.

The remaining tasks are to make sure the unit tests for checkYourAnswers are updated to cover this change in design.

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Are any changes required to the e2e tests?
- [ ] If you've added a new route, have you added a new
  `auditEvent`? (see `server/routes/apply.ts` for examples)
- [ ] Are there environment variables or other infrastructure configuration which needs to be included in this release?
- [ ] Are there any data migrations required. Automatic or manual?
- [ ] Does this rely on changes being deployed to the CAS API?

## Post merge checklist

Once we've merged it will be auto-deployed to the dev environment.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-community-accommodation-tier-2-ui).

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

### CODE REVIEW

File: server/utils/checkYourAnswersUtils.ts [hunk 1]
```diff
@@ -293,3 +298,33 @@ export const getApplicantDetails = (application: Application | Cas2SubmittedAppl
...
+const processFindMatchedKeys = async (jsonBlob: any, keysToFind: string[]): Promise<string[]> => {
+  const keys = Object.keys(jsonBlob)
+  const result = await Promise.all(keys.map(key => findMatchedKeys(jsonBlob[key], keysToFind)))
+  return result.flat()
+}
+
+const filterKeysWithoutQuestions = async (questions: any, task: string, pageKeys: string[]): string[] => {
+  const permanentPagesWithoutQuestions = ['summary-data', 'oasys-import']
+  const dynamicPagesKeys = pageKeys.filter(item => !permanentPagesWithoutQuestions.includes(item))
+  const matchedKeys = await processFindMatchedKeys(questions[task], dynamicPagesKeys)
```
Comment (tahb) [inline server/utils/checkYourAnswersUtils.ts:327]: We fan out across the list of keys here and each task will drill down and compare keys. Rather than going through each key and sub key sequentially.

File: server/utils/checkYourAnswersUtils.test.ts [hunk 2]
```diff
@@ -78,9 +79,12 @@ describe('checkYourAnswersUtils', () => {
 
   describe('getTaskAnswersAsSummaryListItems', () => {
     it('returns an array of summary list items for a given task', () => {
-      jest.spyOn(getQuestionsUtil, 'getQuestions').mockImplementationOnce(jest.fn(() => mockQuestions))
-
-      jest.spyOn(checkYourAnswersUtils, 'getPage').mockReturnValue(jest.fn())
+      const mockQuestionsForOneTask = {
```
Comment (tahb) [inline server/utils/checkYourAnswersUtils.test.ts:82]: We redefine the mock questions with just one task, since that's all we care about in this scenario.

File: server/utils/checkYourAnswersUtils.ts [hunk 3]
```diff
@@ -36,23 +36,26 @@ export const getTaskAnswersAsSummaryListItems = (
 ): Array<SummaryListItem | QuestionAndAnswer> => {
   const items: Array<SummaryListItem | QuestionAndAnswer> = []
 
+  // Get the latest question schema
```
Comment (tahb) [inline server/utils/checkYourAnswersUtils.ts:39]: I try not to add comments but this code felt like variable and method naming alone didn't provide enough.

**Comment by tahb** [comment]
Looks like I've got a failing integration test at the last minute so will still need to hand this over with a tiny bit more to do @froddd @patrickjfl.

The failure is where it can't find "Arson". Presumably the new code is filtering this out as is viewing it as old question.

Screenshot 2024-05-31 at 17 46 36

https://github.com/ministryofjustice/hmpps-community-accommodation-tier-2-ui/blob/b05f338a698f290f3c1974f3bfd40c5b9d0bb1f3/integration_tests/tests/apply/check_your_answers/check_your_answers.cy.ts#L19 points to the applicationData.json [where "arson" appears in both](https://github.com/ministryofjustice/hmpps-community-accommodation-tier-2-ui/blob/b05f338a698f290f3c1974f3bfd40c5b9d0bb1f3/integration_tests/fixtures/applicationData.json#L303) "current-offences" and "offending-history". [The current questions.ts says that arson is not a type of current offence?](https://github.com/ministryofjustice/hmpps-community-accommodation-tier-2-ui/blob/b05f338a698f290f3c1974f3bfd40c5b9d0bb1f3/server/form-pages/utils/questions.ts#L795), is the fix here to remove it from the applicationData.json?

File: server/utils/checkYourAnswersUtils.test.ts (hunk 2 above)
Comment (froddd) [reply server/utils/checkYourAnswersUtils.test.ts:82]: Not entirely sure this makes this test more readable, having the same questions context for the whole test suite seems fairly easy to grok -- effectively this only duplicates a subset of the existing context test data?

**Review by froddd — APPROVED** [review]
Seems sensible and does the job it needs to do, I may feel a slight lack of linear readability but this may well be due to not being familiar with this codebase!

File: server/utils/checkYourAnswersUtils.test.ts (hunk 2 above)
Comment (patrickjfl) [reply server/utils/checkYourAnswersUtils.test.ts:82]: yes, avoiding a long and verbose expectation statement. But happy to defer if you want to set it up differently @froddd

File: server/utils/checkYourAnswersUtils.ts [hunk 4]
```diff
@@ -36,23 +36,26 @@ export const getTaskAnswersAsSummaryListItems = (
...
-        pageKey,
-        questions,
-        outputFormat,
-      })
-    }
+  // Get the page keys stored on the application at creation
+  const applicationPageKeys = getKeysForPages(application, task).filter(key => pagesWeNeverWantToPresent(key) === false)
+
+  // Filter out any keys that are no longer in the latest question schema
+  const relevantPagesKeys = removeAnyOldPageKeys(questions, task, applicationPageKeys)
```
[the remark this answers is not in the document]
Comment (patrickjfl) [reply server/utils/checkYourAnswersUtils.ts:46]: Personally I think it makes more sense to have two functions here, for readability and separation of concerns. We want getKeysForPages to go and get all the page keys for the task from the application data blob. And then we want to go and verify these keys, against an up to date list of the keys that we need. I think doing all that in one function probably makes things slightly harder to follow?

File: server/utils/checkYourAnswersUtils.test.ts (hunk 2 above)
Comment (froddd) [reply server/utils/checkYourAnswersUtils.test.ts:82]: The expectation statement shouldn't change, given the method under test is called with `task1` specifically. Removing `mockQuestionsForOneTask` and passing `mockQuestions` to the tested method tests the same assertion and doesn't require the expectation to change, which is why I find it a bit odd to duplicate it in this case.

Comment (patrickjfl) [reply server/utils/checkYourAnswersUtils.test.ts]: ah I see what you mean! yes, sorry Fred, good point and I agree. I've removed it 👍
